### ISSUE TITLE
Learning rate decay

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#679]
I've read that one should usually use learning rate decay during training, and that's it's really important for getting good results.  But I haven't seen any DeepChem models using it, and given how optimization is implemented in TensorGraph, I don't see any straightforward way to do it.  How can we make this possible?

**Comment by lilleswing** [comment]
@peastman were the learning rate decay papers using Adam ( the default optimizer in TG)? 
Adam already does a decay followed by a raise as described here
https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer
https://arxiv.org/pdf/1412.6980v8.pdf
And most of what I have read is that Adam handles decay on its own.

Having said that we have https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorgraph/tensor_graph.py#L88
the global step variable and we can set a decay schedule in the constructor based on global steps.

**Comment by peastman** [comment]
I don't think that's actually learning rate decay.  See the discussion at https://stats.stackexchange.com/questions/200063/tensorflow-adam-optimizer-with-exponential-decay.  Including the second answer which said the same thing you did, and then someone else commented on it to explain why it's incorrect.

> and we can set a decay schedule in the constructor based on global steps.

Given the current API, I don't see how that's possible.  At least not in a way that doesn't break saving models.

**Comment by rbharath** [comment]
One intermediate option might be to support fixed decay schedules (linear, exponential, etc). We could store these as strings, which would provide no issues with saving.

A general decay schedule would have to be a stored function. Can we pickle functions?

**Comment by patrickhop** [comment]
I think Adam decays the beta terms for the exponential moving averages, not the learning rate alpha.

**Comment by lilleswing** [comment]
https://github.com/deepchem/deepchem/pull/680
Is a stab I took at getting learning rate decay into TG.  It works but I still haven't fully tested whether it is pickle-able and how to save the Optimizer itself (this is less important for my use case of using models for inference).

```python
import matplotlib.pyplot as plt
import math
%matplotlib inline

epsilon=1e-08
beta1=0.9
beta2=0.999
learning_rate = 0.001
```


```python
def lr_t(t):
    return learning_rate * math.sqrt(1.0 - beta2**t + epsilon) / (1.0 - beta1 **t + epsilon)
```


```python
x1 = list(range(10000))[1:]
y1 = [lr_t(x) for x in range(10000)][1:]
```


```python
plt.plot(x1,y1)
```




    [<matplotlib.lines.Line2D at 0x7ff7854a7c50>]




![output_3_1](https://user-images.githubusercontent.com/1225605/28420325-cad09684-6d2f-11e7-8994-4d6fc9dbe7d2.png)
It dips during "warm up" and then regrows to learning_rate


The variable also only changes depending on the magnitude of its first and second derivative

``` bash
m_t <- beta1 * m_{t-1} + (1 - beta1) * g
v_t <- beta2 * v_{t-1} + (1 - beta2) * g * g
variable <- variable - lr_t * m_t / (sqrt(v_t) + epsilon)
```


```python

```

**Comment by peastman** [comment]
I'll go ahead and close this, since #700 created a simple API for it.
